cmd/otel-allocator/prehook/relabel.go

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prehook
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/targetscommon"
+)
+
+type RelabelConfigTargetFilter struct {
+	m          sync.RWMutex
+	log        logr.Logger
+	relabelCfg map[string][]*relabel.Config
+}
+
+func NewRelabelConfigTargetFilter(log logr.Logger) Hook {
+	return &RelabelConfigTargetFilter{
+		log:        log,
+		relabelCfg: make(map[string][]*relabel.Config),
+	}
+}
+
+// helper function converts from model.LabelSet to []labels.Label.
+func ConvertLabelToPromLabelSet(lbls model.LabelSet) []labels.Label {
+	newLabels := make([]labels.Label, len(lbls))
+	index := 0
+	for k, v := range lbls {
+		newLabels[index].Name = string(k)
+		newLabels[index].Value = string(v)
+		index++
+	}
+	return newLabels
+}
+
+func (tf *RelabelConfigTargetFilter) Apply(targets map[string]*targetscommon.TargetItem) map[string]*targetscommon.TargetItem {
```

Comment:
Is there a possibility that `Apply()` might be called before the first `SetConfig()` call, when `relabelCfg` is empty? The resulting behavior of no targets available seems okay on startup, I'm just wondering how long that could last?

Reply:
Hmm not obvious to me that anything is stopping that from occurring. I would expect SetConfig to be called when configmap is created initially which would happen before any targets are discovered.. I added the following at the start of the function so no filtering will occur in this case

```
	if len(tf.relabelCfg) == 0 {
		return targets
	}
```